### PR TITLE
Generate drum od phases core

### PR DESCRIPTION
**Pull request by Purplo-cf**
Static GenerateActivationPhrases method added to InstrumentDifficulty extensions, similar to how the No Kick modifier is handled.

Static DrumsEngine.IsIdealActivationNote method searches potential activation notes for combinations of either Crash+Kick or Crash+Snare in all three drum engine modes. If the note directly on the bar line doesn't pass the test, an additional search will run +/- an 8th note from the bar line for a better candidate. This is for phrases with a heavy accent either right before or right after the start of the next measure.

### CODE REVIEW

**Comment by Purplo-cf** [comment]
I'm using Enumerable.Where to filter the Phrases list into separate lists that just contain StarPower Phrases and Solo Phrases. I kind of remember seeing "LINQ = slow" discussions on the discord, so let me know if you want me to handle this a different way

**Review by EliteAsian123 — CHANGES_REQUESTED** [review]
Hi, sorry for the late review, I've been quite busy.

I provided some feedback below. I have not tested this in-game yet so I'll provide more feedback when I get the chance. The algorithm itself is absolutely massive, so if you could simplify it somehow or at least split it up a bit that'll make it easier to understand.

Thanks!

File: YARG.Core/Chart/Tracks/InstrumentDifficultyExtensions.cs [hunk 1]
```diff
@@ -198,5 +202,211 @@ public static void RemoveKickDrumNotes(this InstrumentDifficulty<DrumNote> diffi
                 }
             }
         }
+
+        public static void GenerateActivationPhrases(this InstrumentDifficulty<DrumNote> difficulty, SyncTrack syncTrack)
```
[the remark this answers is not in the document]
Comment (Purplo-cf) [reply YARG.Core/Chart/Tracks/InstrumentDifficultyExtensions.cs:206]: > This also reminds me: this change isn't necessarily linked to the new engine, so this is fine to be on the `master` branch.

I'm still moving stuff around here, but is it cool if I keep this on the `engine-refactor` branch? I've got a couple references to the new `Note.AllNotes` Enumerator in here, but I can change them back to `ChordEnumerator` if you prefer

https://github.com/YARC-Official/YARG.Core/blob/fe828a74ef7aa9020fd129690a6bda0a60b18b87/YARG.Core/Chart/Tracks/InstrumentDifficultyExtensions.cs#L378-L382

https://github.com/YARC-Official/YARG.Core/blob/fe828a74ef7aa9020fd129690a6bda0a60b18b87/YARG.Core/Engine/Drums/DrumsEngine.cs#L271-L279

[the remark this answers is not in the document]
Comment (EliteAsian123) [reply YARG.Core/Chart/Tracks/InstrumentDifficultyExtensions.cs:206]: `engine-refactor` probably won't be merged into dev for at least a few months, so it's probably for the best to move these into `dev` if possible. Thanks!

**Comment by TheNathannator** [comment]
That branch merge should be a rebase + force-push instead, keeps the Git history for this PR cleaner. You should be able to undo the merge commit and it'll undo the whole merge AFAIK.

**Comment by Purplo-cf** [comment]
Sorry for the mess! After the failed rebase I actually started over on a new master branch. I was hoping to close this once the revisions were ready and open a new PR since this is so early.

I'll try force pushing the new branch first when it's ready and see what happens.

**Comment by Purplo-cf** [comment]
I clearly have a lot to learn about rebasing, but I don't want that to get in the way of the review. I'm cutting my losses here and starting over with a new PR

**Pull request closed without merging**